File: problems/1856_maximum_subarray_min_product/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int maxSumMinProduct(vector<int>& nums) {
        long long MOD = 1e9 + 7;
        int n = nums.size();
        vector<long long> prefix(n + 1, 0);
        for (int i = 0; i < n; i++) prefix[i + 1] = prefix[i] + nums[i];
        vector<int> left(n, -1), right(n, n);
        stack<int> st;
        for (int i = 0; i < n; i++) {
            while (!st.empty() && nums[st.top()] >= nums[i]) { right[st.top()] = i; st.pop(); }
            if (!st.empty()) left[i] = st.top();
            st.push(i);
        }
        long long result = 0;
        for (int i = 0; i < n; i++) {
            long long total = prefix[right[i]] - prefix[left[i] + 1];
            result = max(result, (long long)nums[i] * total);
        }
        return result % MOD;
    }
};

File: problems/1856_maximum_subarray_min_product/Solution.java
import java.util.*;

class Solution {
    public int maxSumMinProduct(int[] nums) {
        long MOD = 1000000007;
        int n = nums.length;
        long[] prefix = new long[n + 1];
        for (int i = 0; i < n; i++) prefix[i + 1] = prefix[i] + nums[i];
        int[] left = new int[n], right = new int[n];
        Arrays.fill(left, -1);
        Arrays.fill(right, n);
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] >= nums[i]) right[stack.pop()] = i;
            if (!stack.isEmpty()) left[i] = stack.peek();
            stack.push(i);
        }
        long result = 0;
        for (int i = 0; i < n; i++) {
            long total = prefix[right[i]] - prefix[left[i] + 1];
            result = Math.max(result, (long) nums[i] * total);
        }
        return (int) (result % MOD);
    }
}
